File: spring-boot-tipps-reactive/pnia/src/main/java/io/github/jlmc/reactive/domain/usecases/GetPhoneNumberAggregation.java
package io.github.jlmc.reactive.domain.usecases;

import io.github.jlmc.reactive.domain.model.PhoneNumber;
import io.github.jlmc.reactive.domain.services.blacklist.BlackListService;
import io.github.jlmc.reactive.domain.services.blacklist.ForbiddenInputValuesException;
import io.github.jlmc.reactive.domain.services.prefixes.PrefixesService;
import io.github.jlmc.reactive.domain.services.sectors.BusinessSectorService;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.stream.Collectors.groupingBy;

@Service
public class GetPhoneNumberAggregation {

    private final PrefixesService prefixesService;
    private final BusinessSectorService businessSectorService;
    private final BlackListService blackListService;

    public GetPhoneNumberAggregation(PrefixesService prefixesService,
                                     BusinessSectorService businessSectorService,
                                     BlackListService blackListService) {
        this.prefixesService = prefixesService;
        this.businessSectorService = businessSectorService;
        this.blackListService = blackListService;
    }

    public Mono<Map<String, Map<String, Long>>> execute(List<String> phoneNumbersAsText) {

        Collection<String> forbiddenInputs = blackListService.getForbiddenInputs(phoneNumbersAsText);
        if (!forbiddenInputs.isEmpty()) {
            throw new ForbiddenInputValuesException(forbiddenInputs);
        }

        Stream<String> nonNullPhoneNumbers = phoneNumbersAsText.stream().filter(Objects::nonNull);

        Flux<PhoneNumber> flux =
                Flux.fromStream(nonNullPhoneNumbers)
                    .filter(PhoneNumber::isValidPhoneNumber)
                    .map(PhoneNumber::of)
                    .flatMap(this::prefixOf)
                    .flatMap(phoneNumber -> businessSectorService.getNumberSector(phoneNumber)
                                                                 .filter(Objects::nonNull)
                                                                 .map(phoneNumber::withSector));

        return flux
                .collect(
                        Collectors.groupingBy(PhoneNumber::getPrefix,
                                Collectors.groupingBy(p -> p.getSector().getName(),
                                        Collectors.counting()))
                );
    }

    private Mono<PhoneNumber> prefixOf(PhoneNumber phoneNumber) {
        return prefixesService.findByPhoneNumber(phoneNumber.getNumber())
                              .filter(Objects::nonNull)
                              .map(phoneNumber::withPrefix);
    }
}

File: pnia-kotlin/src/main/kotlin/io/github/jlmc/pniakotlin/domain/interactors/GetPhoneNumberAggregation.kt
package io.github.jlmc.pniakotlin.domain.interactors

import io.github.jlmc.pniakotlin.domain.model.PhoneNumber
import io.github.jlmc.pniakotlin.domain.services.blacklist.BlackListService
import io.github.jlmc.pniakotlin.domain.services.prefixes.PrefixesService
import io.github.jlmc.pniakotlin.domain.services.sectors.BusinessSectorService
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.stream.Collectors

@Service
class GetPhoneNumberAggregation(
    private val blackListService: BlackListService,
    private val prefixesService: PrefixesService,
    private val businessSectorService: BusinessSectorService,
) {

    fun execute(phoneNumbers: List<String?>): Mono<Map<String, Map<String, Long>>> {
        blackListService.validate(phoneNumbers)

        val flux: Mono<Map<String, Map<String, Long>>> =
            Flux.fromIterable(phoneNumbers)
                .flatMap { it?.let { Mono.just(it) }?: Mono.empty() }
                .filter(PhoneNumber.Companion::isValidPhoneNumber)
                .flatMap { phoneNumberOf(it) }
                .flatMap { fetchPrefix(it) }
                .flatMap { fetchSector(it) }
                .collect(
                    Collectors.groupingBy(
                        PhoneNumber::prefix,
                        Collectors.groupingBy(
                            { it.sector!!.name },
                            Collectors.counting()
                        )
                    )
                )

        return flux
    }

    private fun phoneNumberOf(txt: String?) = txt?.let {
        Mono.fromCallable { PhoneNumber.of(it) }
    } ?: Mono.empty()

    private fun fetchPrefix(phoneNumber: PhoneNumber): Mono<PhoneNumber> {
        return prefixesService.findByPhoneNumber(phoneNumber.number)
            .filter { it != null }
            .map { phoneNumber.withPrefix(it) }
    }

    private fun fetchSector(phoneNumber: PhoneNumber): Mono<PhoneNumber> {
        return businessSectorService.getNumberSector(phoneNumber)
            .filter { it != null }
            .map { phoneNumber.withSector(it) }
    }

}
